python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -1479,94 +1343,84 @@ def _execute_asset_backfill_iteration_inner(
+    materialized_since_last_tick = (
+        updated_materialized_subset - asset_backfill_data.materialized_subset
+    )
+    logger.info(
+        f"Assets materialized since last tick:\n{_asset_graph_subset_to_str(materialized_since_last_tick, asset_graph)}"
+        if not materialized_since_last_tick.is_empty
+        else "No relevant assets materialized since last tick."
+    )
 
-        materialized_since_last_tick = (
-            updated_materialized_subset - asset_backfill_data.materialized_subset
-        )
-        logger.info(
-            f"Assets materialized since last tick:\n{_asset_graph_subset_to_str(materialized_since_last_tick, asset_graph)}"
-            if not materialized_since_last_tick.is_empty
-            else "No relevant assets materialized since last tick."
-        )
+    failed_asset_graph_subset = _get_failed_asset_graph_subset(
+        asset_graph_view,
+        backfill_id,
+        updated_materialized_subset,
+    )
 
-        failed_asset_graph_subset = _get_failed_asset_graph_subset(
-            asset_graph_view,
-            backfill_id,
-            updated_materialized_subset,
-        )
-        candidate_asset_graph_subset = _get_candidate_asset_graph_subset(
-            asset_backfill_data,
-            asset_graph_view,
-            updated_materialized_subset,
-            failed_asset_graph_subset,
-        )
+    failed_and_downstream_subset = _get_failed_and_downstream_asset_graph_subset(
+        backfill_id,
+        asset_backfill_data,
+        asset_graph_view,
+        updated_materialized_subset,
+        failed_asset_graph_subset,
+    )
 
-        failed_and_downstream_subset = _get_failed_and_downstream_asset_graph_subset(
-            backfill_id,
-            asset_backfill_data,
-            asset_graph_view,
-            updated_materialized_subset,
-            failed_asset_graph_subset,
-        )
+    candidate_asset_graph_subset = (
+        asset_backfill_data.target_subset
+        - asset_backfill_data.requested_subset
+        - failed_and_downstream_subset
+    )
 
     logger.info(
         f"Considering the following candidate subset:\n{_asset_graph_subset_to_str(candidate_asset_graph_subset, asset_graph)}"
         if not candidate_asset_graph_subset.is_empty
         else "Candidate subset is empty."
     )
 
-    asset_subset_to_request, not_requested_and_reasons = bfs_filter_asset_graph_view(
-        asset_graph_view,
-        lambda candidate_asset_graph_subset,
-        visited: _should_backfill_atomic_asset_graph_subset_unit(
+    asset_graph_subset_to_request = AssetGraphSubset()
+    not_requested_and_reasons = []
+    for asset_key in asset_graph.toposorted_asset_keys:
+        candidate_subset = asset_graph_view.get_entity_subset_from_asset_graph_subset(
+            candidate_asset_graph_subset, asset_key
+        )
+        if candidate_subset.is_empty:
+            continue
+
+        result = _should_backfill_atomic_asset_graph_subset_unit(
```

Comment:
i think we will definitely want https://app.graphite.dev/github/pr/dagster-io/dagster/31538/ once the candidate set is so much bigger - the bulk of the failure reasons listed after this change are going to be fairly nonsensical otherwise

Reply:
indeed! rebased on this